In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/auto-sales-data/Auto Sales data.csv


In [2]:
# Import necessary libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [3]:
# Creating a Random Forest Regressor Model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model

RandomForestRegressor(random_state=42)

In [4]:
df = pd.read_csv('/kaggle/input/auto-sales-data/Auto Sales data.csv')

In [5]:
df = df.drop(['ORDERDATE','POSTALCODE','CONTACTLASTNAME', 'CONTACTFIRSTNAME','ORDERNUMBER','PHONE', 'ADDRESSLINE1','POSTALCODE'],axis=1)

#df['ORDERDATE'] = pd.to_datetime(df['ORDERDATE'],yearfirst = True ) # changing the date time format to a suitable format

df.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  DAYS_SINCE_LASTORDER  \
0               30      95.70                2  2871.00                   828   
1               34      81.35                5  2765.90                   757   
2               41      94.74                2  3884.34                   703   
3               45      83.26                6  3746.70                   649   
4               36      96.66                1  3479.76                   586   

    STATUS  PRODUCTLINE  MSRP PRODUCTCODE          CUSTOMERNAME        CITY  \
0  Shipped  Motorcycles    95    S10_1678     Land of Toys Inc.         NYC   
1  Shipped  Motorcycles    95    S10_1678    Reims Collectables       Reims   
2  Shipped  Motorcycles    95    S10_1678       Lyon Souveniers       Paris   
3  Shipped  Motorcycles    95    S10_1678     Toys4GrownUps.com    Pasadena   
4  Shipped  Motorcycles    95    S10_1678  Technics Stores Inc.  Burlingame   

  COUNTRY DEALSIZE  
0     USA    Small  
1  France    Small  
2  France   Medium  
3     USA   Medium  
4     USA   Medium

In [6]:
df.nunique()

QUANTITYORDERED           58
PRICEEACH               1843
ORDERLINENUMBER           18
SALES                   2690
DAYS_SINCE_LASTORDER    1832
STATUS                     6
PRODUCTLINE                7
MSRP                      80
PRODUCTCODE              109
CUSTOMERNAME              89
CITY                      71
COUNTRY                   19
DEALSIZE                   3
dtype: int64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2747 entries, 0 to 2746
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   QUANTITYORDERED       2747 non-null   int64  
 1   PRICEEACH             2747 non-null   float64
 2   ORDERLINENUMBER       2747 non-null   int64  
 3   SALES                 2747 non-null   float64
 4   DAYS_SINCE_LASTORDER  2747 non-null   int64  
 5   STATUS                2747 non-null   object 
 6   PRODUCTLINE           2747 non-null   object 
 7   MSRP                  2747 non-null   int64  
 8   PRODUCTCODE           2747 non-null   object 
 9   CUSTOMERNAME          2747 non-null   object 
 10  CITY                  2747 non-null   object 
 11  COUNTRY               2747 non-null   object 
 12  DEALSIZE              2747 non-null   object 
dtypes: float64(2), int64(4), object(7)
memory usage: 279.1+ KB


In [8]:
label_encoder = LabelEncoder()

df['STATUS'] = label_encoder.fit_transform(df['STATUS'])
df['PRODUCTLINE'] = label_encoder.fit_transform(df['PRODUCTLINE'])
df['PRODUCTCODE'] = label_encoder.fit_transform(df['PRODUCTCODE'])
df['CUSTOMERNAME'] = label_encoder.fit_transform(df['CUSTOMERNAME'])
df['CITY'] = label_encoder.fit_transform(df['CITY'])
df['COUNTRY'] = label_encoder.fit_transform(df['COUNTRY'])
df['DEALSIZE'] = label_encoder.fit_transform(df['DEALSIZE'])

df.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  DAYS_SINCE_LASTORDER  \
0               30      95.70                2  2871.00                   828   
1               34      81.35                5  2765.90                   757   
2               41      94.74                2  3884.34                   703   
3               45      83.26                6  3746.70                   649   
4               36      96.66                1  3479.76                   586   

   STATUS  PRODUCTLINE  MSRP  PRODUCTCODE  CUSTOMERNAME  CITY  COUNTRY  \
0       5            1    95            0            44    41       18   
1       5            1    95            0            63    55        6   
2       5            1    95            0            45    51        6   
3       5            1    95            0            82    52       18   
4       5            1    95            0            76    12       18   

   DEALSIZE  
0         2  
1         2  
2         1  
3         1  
4         1

In [9]:
# 'SALES' is the target variable
X = df.drop('SALES', axis=1)  # Features
y = df['SALES']  # Target variable

In [10]:
# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
# Training the model
rf_model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [12]:
# Making predictions on the test set
y_pred = rf_model.predict(X_test)
y_pred

array([ 2398.3337,  2010.2661,  2544.1323,  2482.1258,  1706.0424,
        3990.3014,  2177.6186,  7729.3487,  3206.9595,  5985.8959,
        3155.9774,  2225.5041,  1308.7294,  3303.3774,  3201.3601,
        5642.1716,  6568.3367,  2859.8725,  6771.2117,  2341.7629,
        6815.5982,  2465.6314,  2412.2423,  1768.5654,  1514.9656,
        1245.6318,  3178.6614,  6501.5417,  2799.58  ,  4808.0787,
        3214.3548,  5731.2542,  5517.9562,  4172.2864,  4625.961 ,
        3577.2125,  5130.5928,  3716.4861,  2186.9797,  1867.0614,
        1830.3003,  1240.5828,  2339.2115,  4449.1504,  2832.677 ,
        2351.3359,  3829.1491,  1978.0111,  7592.7269,  3127.6349,
        3257.0017,  1868.2464,  2201.3928,  1363.2772,  6029.4201,
        2341.0308,  6600.5131,  6231.4333,  1631.3763,  1693.6582,
        3895.4597,  6527.3612,  3475.7944,  4164.8367,  6034.9297,
        7425.4305,  1534.0231,  2829.7143,  3987.3238,  2809.1194,
        1627.8147,  6421.0542,  2155.1817,  1632.2048,  6585.3

In [13]:
# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

Mean Absolute Error: 69.24167672727283
Mean Squared Error: 20671.68980515293
R-squared: 0.9932187078509546


**Hyperparameter Tuning**

In [14]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint

In [15]:
# Defining hyperparameter grid for tuning
param_dist = {
    'n_estimators': randint(10, 100),
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

In [16]:
# Using Randomized Search to find the best hyperparameters
random_search = RandomizedSearchCV(estimator=rf_model, param_distributions=param_dist, n_iter=10, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)
random_search

RandomizedSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42),
                   n_jobs=-1,
                   param_distributions={'max_depth': [None, 10, 20, 30],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7aa6de471150>},
                   scoring='neg_mean_squared_error')

In [17]:
# Fitting the grid search to the data
random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42),
                   n_jobs=-1,
                   param_distributions={'max_depth': [None, 10, 20, 30],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7aa6de471150>},
                   scoring='neg_mean_squared_error')

In [18]:
# Getting the best hyperparameters from the grid search
best_params = random_search.best_params_

best_params

{'max_depth': None,
 'min_samples_leaf': 2,
 'min_samples_split': 2,
 'n_estimators': 73}

In [19]:
# Creating a Random Forest Regressor model with the best hyperparameters
best_rf_model = RandomForestRegressor(**best_params)

best_rf_model

RandomForestRegressor(min_samples_leaf=2, n_estimators=73)

In [20]:
# Training the model
best_rf_model.fit(X_train, y_train)

RandomForestRegressor(min_samples_leaf=2, n_estimators=73)

In [21]:
# Make predictions on the test set
y_pred = best_rf_model.predict(X_test)

#y_pred

In [22]:
# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Best Hyperparameters: {best_params}')
print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

Best Hyperparameters: {'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 73}
Mean Absolute Error: 70.64978507631292
Mean Squared Error: 21333.61237699632
R-squared: 0.9930015659345449


**Before Hyperparameter Tuning**

Mean Absolute Error: 69.24167672727283

Mean Squared Error: 20671.68980515293

R-squared: 0.9932187078509546


**After Hyperparameter Tuning**

Best Hyperparameters: {'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 57}

Mean Absolute Error: 67.91102291130366

Mean Squared Error: 20985.857879513846

R-squared: 0.9931156458605597